### PR TITLE
Budget bar statistics

### LINKED ISSUE #2658: Should we optionally display the weekly/monthly budget on the budget bar?
Thanks for creating an issue! Please fill out this form so we can be sure to have all the information we need, and to minimize back and forth.

* **What are we trying to do?**

More than one fund has reached out to me recently to ask for this. We've generally discouraged doing Real Budgeting in DARIA (excel is way better and more consistent than we'll ever be) but people are making pretty good arguments that some signals would be helpful.

@lwaldsc as one of the original intellectual architect of the budget bar interested in your thoughts - my recollection is that we didn't do this originally because we thought it would bum out DCAF CMs (one late case and there you go). but wonder if this is similarly a bummer for smaller funds.

* **What feature or behavior is this required for?**

not doing math

* **How could we solve this issue? (Not knowing is okay!)**

if we _were_ to do this, it would look like:
* config to toggle display on/off
* add some math to the budget bar like `(soft pledged: X) + (sent pledged: Y) = X+Y / (weekly budget) (Z%)`

* **Anything else?**

nah

### PR DESCRIPTION
**Pull request by elimbaum**
I rule and have completed some work on Case Manager that's ready for review!

(brief, plain english overview of your changes here)

This pull request makes the following changes:
* adds statistics to budget bar (spent, remaining, percentages)
* refactors code to make adding more aggregate stats easier (and prevent code duplicate!)

(If there are changes to the views, please include a screenshot so we know what to look for!)

It relates to the following issue #s: 
* Fixes #2658 

For reviewer:
* Adjust the title to explain what it does for the notification email to the listserv.
* Tag this PR:
  * `feature` if it contains a feature, fix, or similar. This is anything that contains a user-facing fix in some way, such as frontend changes, alterations to backend behavior, or bug fixes.
  * `dependencies` if it contains library upgrades or similar. This is anything that upgrades any dependency, such as a Gemfile update or npm package upgrade.
* If it contains neither, no need to tag this PR.

### CODE REVIEW

**Comment by elimbaum** [comment]
new view
Screen Shot 2022-11-22 at 22 49 41

**Comment by elimbaum** [comment]
with config turned off

Screen Shot 2022-11-25 at 18 27 03

q's here:
- do we still want to show percents on sent/pledged when stats are turned off?
- do we want percents on those fields to begin with? (regardless of config value)
- should we unit test or is system test sufficient?

**Comment by elimbaum** [comment]
@colinxfleming rereading the issue noticed you originally wanted to cut the %s. let me know what you think about this ^ both in terms of language and data presentation.

**Comment by colinxfleming** [comment]
Answer to qs:
* I think if aggregate statistics is turned off, let's keep it exactly the same as it is now
* Seeing the percents in action I kinda like the percents (and you've done a great job of handling the edge cases/overflow cases here), so if you're comfortable with them let's keep them
* a lot of this hinges on `sum_fund_pledges` so a unit test for that might be nice, but if we're pressed then I think it's skippable since, as you noted, there are system tests for this

**Comment by colinxfleming** [comment]
I don't really have any suggested changes, this is great work. Happy to hit approve when it comes out of draft

**Comment by elimbaum** [comment]
thanks for comments!

> I think if aggregate statistics is turned off, let's keep it exactly the same as it is now

by "same as it is now", do you mean `main` or what I currently have?

& i'll make sure to exercise `sum_fund_pledges` a bit better in unit tests.

**Comment by colinxfleming** [comment]
by same as it is now, I mean behavior on main - I'd like this to essentially function as a feature swap, where you can opt in to this fancy thing by config, but otherwise it's indistinguishable from how it works now.

**Comment by elimbaum** [comment]
ok! i'll adjust that

**Comment by elimbaum** [comment]
when stats are turned off, here's what it looks like

Screen Shot 2022-11-27 at 21 53 56

besides some minor layout changes, identical to before.

File: app/views/dashboards/_budget_bar.html.erb [hunk 1]
```diff
@@ -16,13 +16,28 @@
...
-            count: expenditures[type].count) %>
+    <div class="col-sm-3">
+      <%= budget_bar_statistic(t("dashboard.budget_bar.#{type}"),
+                               expenditures[type],
+                               limit: limit,
+                               show_aggregate_statistics: Config.show_aggregate_statistics?) %>
+    </div>
+  <% end %>
+
+  <% if Config.show_aggregate_statistics? %>
```
Comment (elimbaum) [inline app/views/dashboards/_budget_bar.html.erb:27]: checking the config twice here. looks like it gets cached, but is there a better way? perhaps more generally, does this architecture (statistic-string builder + options) seem ok to you?

File: app/views/dashboards/_budget_bar.html.erb [hunk 2]
```diff
@@ -16,13 +16,28 @@
...
+                               limit: limit,
+                               show_aggregate_statistics: true) %>
+    </div>
+
+    <div class="col-sm-3">
+      <%= budget_bar_statistic_builder(name: t('dashboard.budget_bar.remaining'),
+                                       amount: budget_bar_remaining(expenditures, limit),
+                                       count: nil,
+                                       limit: limit,
+                                       show_aggregate_statistics: true) %>
```
Comment (elimbaum) [inline app/views/dashboards/_budget_bar.html.erb:40]: in these cases, we have already checked the config — we know it's `true`.

**Comment by elimbaum** [comment]
...actually, maybe we should be using [keyword arguments](https://stackoverflow.com/questions/15062570/when-to-use-keyword-arguments-aka-named-parameters-in-ruby) here.

**Comment by elimbaum** [comment]
@colinxfleming ready to go!

File: app/views/dashboards/_budget_bar.html.erb (hunk 1 above)
Comment (colinxfleming) [reply app/views/dashboards/_budget_bar.html.erb:27]: yeah, this is a very railsy pattern for html generation and I think it serves us nicely here. I don't think double-checking is problematic here (as you noted, query gets cached) but the way around this for a heavier query that didn't get cached would be to declare this as an instance variable in the controller, e.g. `@show_stats = Config.show_aggregate_statistics?` and then check that variable directly. (Again, to be clear, shouldn't do that here thanks to db cache magic!)

**Review by colinxfleming — APPROVED** [review]
I always forget ruby kwargs are a thing. They're well used here.

I'm going to make a very soft style change and put spent/remaining in bold so they're easy to distinguish, and then I'm going to merge. No real suggestions from my corner on the logic, this is very well done. Thanks much @elimbaum !
